mllib/src/main/scala/org/apache/spark/mllib/linalg/distributed/BlockMatrix.scala

```diff
@@ -0,0 +1,215 @@
+ * @param numColBlocks Number of blocks that form the columns of the matrix.
+ * @param rowsPerBlock Number of rows that make up each block.
+ * @param colsPerBlock Number of columns that make up each block.
+ */
+private[mllib] class GridPartitioner(
+    val numRowBlocks: Int,
+    val numColBlocks: Int,
+    val rowsPerBlock: Int,
+    val colsPerBlock: Int,
+    override val numPartitions: Int) extends Partitioner {
+
+  /**
+   * Returns the index of the partition the SubMatrix belongs to.
+   *
+   * @param key The key for the SubMatrix. Can be its position in the grid (its column major index)
+   *            or a tuple of three integers that are the final row index after the multiplication,
+   *            the index of the block to multiply with, and the final column index after the
+   *            multiplication.
+   * @return The index of the partition, which the SubMatrix belongs to.
+   */
+  override def getPartition(key: Any): Int = {
+    key match {
+      case (rowIndex: Int, colIndex: Int) =>
+        Utils.nonNegativeMod(rowIndex + colIndex * numRowBlocks, numPartitions)
+      case (rowIndex: Int, innerIndex: Int, colIndex: Int) =>
+        Utils.nonNegativeMod(rowIndex + colIndex * numRowBlocks, numPartitions)
+      case _ =>
+        throw new IllegalArgumentException("Unrecognized key")
+    }
+  }
+
+  /** Checks whether the partitioners have the same characteristics */
+  override def equals(obj: Any): Boolean = {
+    obj match {
+      case r: GridPartitioner =>
+        (this.numPartitions == r.numPartitions) && (this.rowsPerBlock == r.rowsPerBlock) &&
+          (this.colsPerBlock == r.colsPerBlock)
+      case _ =>
+        false
+    }
+  }
+}
+
+/**
+ * Represents a distributed matrix in blocks of local matrices.
+ *
+ * @param numRowBlocks Number of blocks that form the rows of this matrix
+ * @param numColBlocks Number of blocks that form the columns of this matrix
+ * @param rdd The RDD of SubMatrices (local matrices) that form this matrix
+ */
+class BlockMatrix(
+    val numRowBlocks: Int,
+    val numColBlocks: Int,
+    val rdd: RDD[((Int, Int), Matrix)]) extends DistributedMatrix with Logging {
+
+  type SubMatrix = ((Int, Int), Matrix) // ((blockRowIndex, blockColIndex), matrix)
+
+  /**
+   * Alternate constructor for BlockMatrix without the input of a partitioner. Will use a Grid
+   * Partitioner by default.
+   *
+   * @param numRowBlocks Number of blocks that form the rows of this matrix
+   * @param numColBlocks Number of blocks that form the columns of this matrix
+   * @param rdd The RDD of SubMatrices (local matrices) that form this matrix
+   * @param rowsPerBlock Number of rows that make up each block.
+   * @param colsPerBlock Number of columns that make up each block.
+   */
+  def this(
+      numRowBlocks: Int,
+      numColBlocks: Int,
+      rdd: RDD[((Int, Int), Matrix)],
+      rowsPerBlock: Int,
+      colsPerBlock: Int) = {
```

Comment:
The list of arguments cannot provide the complete info about the matrix. For example, if the last block row and the last block column are all missing. Then you cannot figure out the exact matrix size from this list of arguments.

It would be necessary to have `numRows`, `numCols`, `rowsPerBlock`, `colsPerBlock`, and the RDD as input. We can provide factory methods (in follow-up PRs) to create block matrices from other formats, which could figure out the exact `numRows` and `numCols` and use them in the constructor.